src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -1090,23 +1095,99 @@ bool G1CollectedHeap::expand_single_region(uint node_index) {
   return true;
 }
 
-void G1CollectedHeap::shrink_helper(size_t shrink_bytes) {
+void G1CollectedHeap::shrink_with_time_based_selection(size_t shrink_bytes) {
+  if (capacity() == min_capacity()) {
+    log_debug(gc, ergo, heap)("Time-based shrink: Did not shrink the heap (heap already at minimum)");
+    return;
+  }
+
+  size_t aligned_shrink_bytes = os::align_down_vm_page_size(shrink_bytes);
+  aligned_shrink_bytes = align_down(aligned_shrink_bytes, G1HeapRegion::GrainBytes);
+
+  aligned_shrink_bytes = capacity() - MAX2(capacity() - aligned_shrink_bytes, min_capacity());
+  assert(is_aligned(aligned_shrink_bytes, G1HeapRegion::GrainBytes), "Bytes to shrink %zuB not aligned", aligned_shrink_bytes);
+
+  log_debug(gc, ergo, heap)("Time-based shrink: Requested shrink amount: %zuB aligned shrink amount: %zuB",
+                            shrink_bytes, aligned_shrink_bytes);
+
+  if (aligned_shrink_bytes == 0) {
+    log_debug(gc, ergo, heap)("Time-based shrink: Did not shrink the heap (shrink request too small)");
+    return;
+  }
+
+  _verifier->verify_region_sets_optional();
+
+  // We should only reach here from the service thread during idle time
+  // but ensure any GC alloc regions are abandoned
+  _allocator->abandon_gc_alloc_regions();
```

Comment:
Maybe check that they are abandoned (i.e. there are no gc alloc regions) in an assert instead of doing work.